src/Client/ConnectionEstablisher.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <Common/Epoll.h>
+#include <Common/Fiber.h>
+#include <Common/FiberStack.h>
+#include <Common/TimerDescriptor.h>
+#include <Common/PoolWithFailoverBase.h>
+#include <Client/ConnectionPool.h>
+
+namespace DB
+{
+
+/// Class for nonblocking establishing connection to the replica.
+/// It runs establishing connection process in fiber and sets special
+/// read callback which is called when reading from socket blocks.
+/// When read callback is called, socket and receive timeout are added in epoll
+/// and execution returns to the main program.
+/// So, you can poll this epoll file descriptor to determine when to resume.
+class ConnectionEstablisher
+{
+public:
+    using TryResult = PoolWithFailoverBase<IConnectionPool>::TryResult;
+
+    ConnectionEstablisher(IConnectionPool * pool_,
+                          const ConnectionTimeouts * timeouts_,
+                          const Settings * settings_,
+                          const QualifiedTableName * table_to_check = nullptr);
+
+    /// Establish connection with replica, call async_callbeck when
+    /// reading from socket blocks.
+    void establishConnection(AsyncCallback async_callback = {});
+
+    /// In the first call create fiber with establishConnection function,
+    /// in the next - check timeout and resume fiber.
+    void resume();
```

Comment:
Maybe it is reasonable to return fd from resume result.
Or maybe even `variant<fd, fail_message, TryResult>`.
(current interface is also ok)